### PR TITLE
feat(semantic): check optional parameters

### PR DESCRIPTION
**Pull request by Dunqing**

### CODE REVIEW

**Comment by Dunqing** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #2263** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/2263?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/2263?utm_source=stack-comment).

File: tasks/coverage/parser_babel.snap [hunk 1]
```diff
@@ -187,6 +183,22 @@ Expect to Parse: "typescript/class/constructor-with-modifier-names/input.ts"
  4 │ }
    ╰────
 
+Expect to Parse: "typescript/class/parameter-properties/input.ts"
+  × A required parameter cannot follow an optional parameter.
+   ╭─[typescript/class/parameter-properties/input.ts:6:1]
+ 6 │         private pi?: number,
+ 7 │         public readonly pur,
```
Comment (Boshen) [inline tasks/coverage/parser_babel.snap:190]: False positive?

File: tasks/coverage/parser_babel.snap [hunk 2]
```diff
@@ -187,6 +183,22 @@ Expect to Parse: "typescript/class/constructor-with-modifier-names/input.ts"
  4 │ }
    ╰────
 
+Expect to Parse: "typescript/class/parameter-properties/input.ts"
```
Comment (Dunqing) [inline tasks/coverage/parser_babel.snap:186]: In this case, report errors are expected. I tested it in typescript playground the result was correct.

File: tasks/coverage/parser_babel.snap [hunk 3]
```diff
@@ -187,6 +183,22 @@ Expect to Parse: "typescript/class/constructor-with-modifier-names/input.ts"
...
+  × A required parameter cannot follow an optional parameter.
+   ╭─[typescript/class/parameter-properties/input.ts:6:1]
+ 6 │         private pi?: number,
+ 7 │         public readonly pur,
+   ·         ───────────────────
+ 8 │         // Also works on AssignmentPattern
+   ╰────
+
+Expect to Parse: "typescript/function/declare-pattern-parameters/input.ts"
```
Comment (Dunqing) [inline tasks/coverage/parser_babel.snap:195]: Same as above

File: tasks/coverage/parser_babel.snap (hunk 1 above)
Comment (Dunqing) [reply tasks/coverage/parser_babel.snap:190]: Yes, I commented https://github.com/oxc-project/oxc/pull/2263#discussion_r1475656546

File: tasks/coverage/parser_babel.snap [hunk 4]
```diff
@@ -187,6 +183,22 @@ Expect to Parse: "typescript/class/constructor-with-modifier-names/input.ts"
...
+  × A required parameter cannot follow an optional parameter.
+   ╭─[typescript/function/declare-pattern-parameters/input.ts:1:1]
+ 1 │ declare function f([]?, {})
+   ·                         ──
```
Comment (Boshen) [inline tasks/coverage/parser_babel.snap:199]: Why is the span on `{}`

**Comment by Boshen** [comment]
### Merge activity

* **Feb 2, 2:53 AM EST**: @Boshen started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/oxc-project/oxc/2263).

**Pull request merged**